deploy/otakaro/__main__.py

```diff
@@ -0,0 +1,242 @@
+"""An AWS Python Pulumi program"""
+
+import pulumi
+import pulumi_aws as aws
+import pulumi_awsx as awsx
+from alb import ApplicationLoadBalancer, LoadBalancerTarget
+from backend import DigitalTwinBackend
+from certs import AWSCertificate
+from cloudfront import CloudFront
+from efs import EFSVolume
+from fargate import FargateApplication
+from postgres import PostgresInstance
+
+config = pulumi.Config()
+
+# Define Constants
+
+# Cloud resources should be prefixed with this value unless it doesn't make sense to
+# (i.e Route53 DNS entries)
+prefix = "otakaro-dt"
+
+# Networking settings changing these will almost certainly result in a broken deployment
+# To change these, a `pulumi destroy` should be performed, then the change made, and then a new
+# `pulumi up` ran.
+network_cidr_block = "10.0.0.0/16"
+availability_zones = 2
+
+# The code below creates secrets with these names, and assigns them to the containers.
+# If you wish to change these values, go into the Parameter Store in the AWS console and change
+# the values.
+secret_names = config.get_object("secretParameterNames")
+dns_zone_name = "digitaltwins.nz"
+domains = [
+    "otakaro.digitaltwins.nz",
+    "api.otakaro.digitaltwins.nz",
+    "gs.otakaro.digitaltwins.nz",
+]
+aws_region = "ap-southeast-2"
+
+secret_parameters = {
+    secret: aws.ssm.Parameter(
+        secret,
+        name=secret,
+        type="SecureString",
+        value="placeholder",
+        opts=pulumi.ResourceOptions(ignore_changes=["value", "type"]),
+    )
+    for secret in secret_names
+}
+
+
+# Setup DNS Zone
+dns_zone = aws.route53.Zone(
+    dns_zone_name,
+    name=dns_zone_name,
+)
+
+# Create HTTPS Certs
+certs = AWSCertificate(
+    domains=domains,
+    regions=[aws_region, "us-east-1"],
+    zone=dns_zone,
+)
+
+# Define a virtual private cloud (A private network with automatic subnetting)
+vpc = awsx.ec2.Vpc(
+    f"{prefix}-vpc",
+    cidr_block=network_cidr_block,
+    tags={
+        "Name": f"{prefix}-vpc",
+    },
+    enable_dns_hostnames=True,
+    number_of_availability_zones=availability_zones,  # We don't need multiple?
+)
+
+
+# Create a Postgres Instance
+postgres = PostgresInstance(
+    f"{prefix}-pg",
+    vpc=vpc,
+    database_name="otakaro",
+    username="otakaro",
+    password=config.get_secret("databasePassword"),
+    cidr_blocks=[
+        network_cidr_block,
+    ],
+)
+
+# Set up shared filesystems.
+shared_vol = EFSVolume(
+    f"{prefix}-shared-data",
+    vpc=vpc,
+)
+
+
+# Create the ECS Cluster
+cluster = aws.ecs.Cluster(
+    f"{prefix}-ecs",
+    name=f"{prefix}-ecs",
+)
+
+
+# Setup TerriaMap
+terria = FargateApplication(
+    f"{prefix}-terriamap",
+    cluster,
+    vpc,
+    service_port=3001,
+    image_tag="2024.11.22-4",
```

Comment:
Question: Is it right that these are just the initialisation image tags, and that they are replaced with the newest tag by the lambda `updater.py`? So if I were to run `pulumi up`, this would regress back to `"2024.11.22-4`, and then I would have push my images again with a new name?

Not a suggestion for a fix, just a question I want to understand

Reply:
You're correct. Pulumi has the functionality to ignore changes on resources, _but_ `awsx.ecs.FargateService` is a `ComponentResource` object, and trying to ignore on `taskDefinitionArgs.container.image` (conversion to camelCase is required for ignored changes) does not work.

I suspect it might be possible to do it by creating the `TaskDefinition` separately, but that's also a Crosswalk resource and probably a `ComponentResource` object so I opted not to spend my effort going down a rabbithole.